### PR TITLE
optimizations

### PR DESCRIPTION
**Pull request by pca006132**
1. Make `AssignNormals` a template, so we can optimize away one branch.
2. Add "refinement" for collider to check for collision more precisely, so we get less collision pairs, and less work has to be done in `Boolean3`.

### CODE REVIEW

**Comment by pca006132** [comment]
OK, the edge-triangle distance calculation is actually pretty trivial, and seems to fix the issue we had previously. I guess the previous issue is just some precision issue. However, I'm thinking if we should also have some error margin to avoid numerical errors from affecting manifoldness guarantee.

**Comment by pca006132** [comment]
Testing this with openscad shows that it is still not robust... need a bit more investigation.

**Comment by pca006132** [comment]
In my testing, it seems that the computed squared distance can be about 1e7 times larger than precision squared and still has to be added to the collision pairs. I have verified that sliver triangle is not an issue here (which may throw off the normal calculation...), and it doesn't help much even if I use fma to improve the precision of calculation.

**Comment by pca006132** [comment]
Ah, it works when I use long double (80bit float), so apparently it is still some precision issue, but not sure where is causing such a precision loss.

**Comment by pca006132** [comment]
OK, cross products and dot products are getting catastrophic cancellation in some cases...

File: src/utilities/include/manifold/tri_dist.h [hunk 1]
```diff
@@ -82,6 +82,53 @@ inline void EdgeEdgeDist(vec3& x, vec3& y,  // closest points
   y = q + b * u;
 }
 
+inline double EdgeTriangleDistanceSquared(const vec3 p, const vec3 q,
+                                          const std::array<vec3, 3>& abc) {
+  // precompute normals and edge plane perpendicular to the triangle plane
+  vec3 normal = glm::normalize(glm::cross(abc[1] - abc[0], abc[2] - abc[1]));
```
Comment (elalish) [inline src/utilities/include/manifold/tri_dist.h:88]: This can be NaN, or wildly inaccurate even if finite.

**Review by elalish — COMMENTED** [review]
The Tri-edge intersection part of this is going to be very hard. This is breaking the core guarantee of how our Boolean operates, unless we can completely prove that this approximation is always conservative compared to our symbolic-perturbation shadowing answer. We can already tell that our test set is not adequate to verify this.

**Comment by pca006132** [comment]
Indeed, eventually I resorted to interval arithmetic to try to make this test robust, not sure if there are other ways. Haven't yet change the edge-edge distance computation.

**Comment by pca006132** [comment]
Thinking about it, though, I should probably account for floating point operation errors when doing the interval arithmetic? Maybe I should just copy a robust interval arithmetic library in.

**Comment by neilpa** [comment]
As someone who's been mildly interested in robust comp-geom for a while, anything that computes distance is going to be hard and/or slow to do robustly.

If the goal is to remove more false positive triangle pairs early, there are conservative approaches that rely only on plane orientation predicates. Those you can do robustly in floating point and relatively fast with something like [Shewchuk's robust predicates](http://www.cs.cmu.edu/~quake/robust.html).

In this case, you could use one triangle to define the plane, then check the three points of the other triangle. If the orientation of all three points have the same sign, then the triangles can't intersect.

**Comment by elalish** [comment]
> Thinking about it, though, I should probably account for floating point operation errors when doing the interval arithmetic? Maybe I should just copy a robust interval arithmetic library in.

The more fundamental problem is that I'm not sure how we could make the mathematical guarantee. Keep in mind, being exact isn't good enough either - it has to be conservative relative to our other imprecise (but self-consistent) method of finding intersections. 

I think what you're doing fundamentally is collapsing our `s11` and `s02` calculations into this single loop, trading reduced memory for some duplicate calculation. I actually did a similar thing some years ago when I removed the `s01` vectors, and it definitely improved speed. 

I think if we want to do this optimization robustly, we'll have to take that approach, which is a larger refactor. It would mean collapsing all the lines from `Filter11` to `Winding03` down into the collision functions.

**Comment by pca006132** [comment]
I see, I was thinking if we can guarantee that all pairs where the distance is less than epsilon times some factor, we can over-approximate the set of inputs that our s11 and s02 calculation consider as overlapping, but this is probably not as simple and hard to reason about.

**Comment by pca006132** [comment]
Thinking about it, is it possible to establish a bound, in which the intersection checks (s11 and s02) can only accept things closer than some epsilon, and everything further apart must be rejected?

With that, we can make the overapproximation here return true for everything closer than that epsilon. This should not be hard because we can use interval arithmetic, and account for floating point errors very conservatively by assuming the max error for each operation.

**Comment by pca006132** [comment]
btw converting to draft to avoid having that green button that I always wanted to click :)

**Comment by elalish** [comment]
>Thinking about it, is it possible to establish a bound, in which the intersection checks (s11 and s02) can only accept things closer than some epsilon, and everything further apart must be rejected?

I'm just not sure - keep in mind that the winding numbers of the vertices are also fundamentally related to these calculations and must also be consistent. I would need a proof, and I'm not sure how to write one for this. In any case, if the goal is optimization, I think the approach I outlined above will likely give even better results and won't break our existing proof. But something that large we should probably push till after v3.0.

**Comment by pca006132** [comment]
Yeah, we can take this slow. This should not affect v3.0.

**Comment by elalish** [comment]
For now, would you like to reduce this to just the `AssignNormals` change and we can merge that?

**Comment by pca006132** [comment]
Sure

File: test/hull_test.cpp [hunk 2]
```diff
@@ -81,15 +81,6 @@ TEST(Hull, Tictac) {
   EXPECT_NEAR(sphere.NumVert() + tictacSeg, tictac.NumVert(), 1);
 }
 
-#ifdef MANIFOLD_EXPORT
```
Comment (elalish) [inline test/hull_test.cpp:84]: I think it needs this `#ifdef` since it's using Assimp for ReadMesh.

**Review by elalish — CHANGES_REQUESTED** [review]

File: test/hull_test.cpp (hunk 2 above)
Comment (pca006132) [reply test/hull_test.cpp:84]: No, the last few `BooleanComplex` tests are wrapped inside a large `#ifdef MANIFOLD_EXPORT` block.

Comment (elalish) [reply test/hull_test.cpp:84]: Oh, got it, thanks!

**Comment by pca006132** [comment]
I think it is something related to draft PR. I can open a new one.

**Pull request closed without merging**